File: api/routes/screener.py
"""Screener route — runs the multi-agent pipeline on demand."""
from __future__ import annotations

import csv
import io
import json
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import StreamingResponse

from ..auth import get_current_user_optional
from ..db import get_conn
from ..models import ScreenerRequest, ScreenerResponse, User
from ..security import EXPENSIVE_LIMIT, limiter

router = APIRouter(prefix="/api/screener", tags=["screener"])

# Hard cap — Vercel Hobby has a 10-second function timeout. ~50 tickers is the
# largest set the live pipeline can complete in time. Larger universes go
# through the cron-cached `/cached` endpoint.
MAX_LIVE_TICKERS = 50


def _run_pipeline(
    tickers: Optional[List[str]],
    filters: Optional[Dict[str, Any]],
    agent_names: Optional[List[str]],
) -> Dict[str, Any]:
    # Imported lazily so module import (and tests that mock the universe) stay cheap.
    from screener.benchmarks import fetch_benchmarks, market_regime
    from screener.data_fetcher import fetch_batch
    from screener.filters import apply_filters
    from screener.metrics import compute_metrics
    from screener.orchestrator import score_records
    from screener.universe import get_sp500_tickers

    universe = [t.upper() for t in tickers] if tickers else get_sp500_tickers()

    benchmarks = fetch_benchmarks()
    regime = market_regime(benchmarks)

    raw = fetch_batch(universe)
    records: List[Dict[str, Any]] = []
    for ticker, df in raw.items():
        rec = compute_metrics(ticker, df)
        if rec is not None:
            records.append(rec)

    scored_df = score_records(records, benchmarks=benchmarks, regime=regime, agent_names=agent_names)
    if filters:
        # apply_filters takes records, not a DataFrame
        scored_records = scored_df.to_dict("records") if not scored_df.empty else []
        scored_df = apply_filters(scored_records, filters)

    results = scored_df.to_dict("records") if not scored_df.empty else []
    return {
        "regime": regime,
        "benchmarks": benchmarks,
        "results": results,
        "ran_at": datetime.now(timezone.utc),
    }


@router.post("/run", response_model=ScreenerResponse)
@limiter.limit(EXPENSIVE_LIMIT)
def run_screener(
    request: Request,
    payload: ScreenerRequest,
    user: User = Depends(get_current_user_optional),
) -> ScreenerResponse:
    """Run live screener on a custom ticker list (≤ 50 tickers)."""
    if payload.tickers and len(payload.tickers) > MAX_LIVE_TICKERS:
        raise HTTPException(
            status_code=413,
            detail=(
                f"Live screener accepts up to {MAX_LIVE_TICKERS} tickers. "
                f"Use GET /api/screener/cached for the full S&P 500."
            ),
        )
    out = _run_pipeline(payload.tickers, payload.filters, payload.agents)
    return ScreenerResponse(**out)


@router.get("/cached", response_model=ScreenerResponse)
def get_cached_screener(
    _user: Optional[User] = Depends(get_current_user_optional),
) -> ScreenerResponse:
    """Return latest pre-computed screener results (updated every 4 hours via cron).

    Available to all users (Free + Pro). Returns instantly from Postgres cache.
    Never hits yfinance live.
    """
    try:
        with get_conn() as c:
            with c.cursor() as cur:
                cur.execute(
                    """SELECT ran_at, payload FROM screener_results
                       ORDER BY ran_at DESC LIMIT 1"""
                )
                row = cur.fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="No cached results yet. Cron will populate them.")
        ran_at_str, payload_str = row
        payload = json.loads(payload_str)
        return ScreenerResponse(
            regime={},
            benchmarks={},
            results=payload.get("results", []),
            ran_at=ran_at_str,
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch cached results: {e}")


@router.get("/top")
def get_top_screener(
    n: int = 10,
    _user: Optional[User] = Depends(get_current_user_optional),
) -> list:
    """Return the top N leaders from cached pre-computed results.

    Returns a minimal list of dicts: {ticker, ret_7d, score, best_strategy}.
    Useful for integrations, dashboards, and embedding in digest emails.
    """
    if n < 1 or n > 100:
        raise HTTPException(status_code=400, detail="n must be between 1 and 100")

    try:
        with get_conn() as c:
            with c.cursor() as cur:
                cur.execute(
                    """SELECT payload FROM screener_results
                       ORDER BY ran_at DESC LIMIT 1"""
                )
                row = cur.fetchone()
        if not row:
            return []
        payload = json.loads(row[0])
        results = payload.get("results", [])
        return [
            {
                "ticker": r.get("ticker"),
                "ret_7d": r.get("ret_7d"),
                "score": r.get("score") or r.get("composite_score"),
                "best_strategy": r.get("best_strategy"),
                "vs_voo": r.get("vs_voo"),
            }
            for r in results[:n]
        ]
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch top results: {e}")


@router.get("/export")
def export_screener_csv(
    user: User = Depends(get_current_user_optional),
) -> StreamingResponse:
    """Download screener results as CSV (Pro only).

    Returns the latest cached screener run as a downloadable CSV file with
    all scored columns. Suitable for import into Excel / Google Sheets.
    """
    try:
        with get_conn() as c:
            with c.cursor() as cur:
                cur.execute(
                    """SELECT ran_at, payload FROM screener_results
                       ORDER BY ran_at DESC LIMIT 1"""
                )
                row = cur.fetchone()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"DB error: {e}")

    if not row:
        raise HTTPException(status_code=404, detail="No screener results to export yet.")

    ran_at, payload_str = row
    payload = json.loads(payload_str)
    results = payload.get("results", [])

    if not results:
        raise HTTPException(status_code=404, detail="No results in latest screener run.")

    # Build CSV in memory
    output = io.StringIO()
    fieldnames = [
        "ticker", "price", "ret_7d", "change_7d", "score", "composite_score",
        "momentum", "breakout", "volume", "rs", "mean_reversion",
        "best_strategy", "vs_voo",
    ]
    # Include any extra columns present in the data
    extra_keys = sorted(
        k for k in results[0].keys()
        if k not in fieldnames and not k.startswith("_")
    )
    all_fields = fieldnames + extra_keys

    writer = csv.DictWriter(output, fieldnames=all_fields, extrasaction="ignore")
    writer.writeheader()
    writer.writerows(results)

    filename_date = ran_at.strftime("%Y%m%d") if ran_at else "latest"
    filename = f"best7daysmula_screener_{filename_date}.csv"

    return StreamingResponse(
        iter([output.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.get("/history")
def screener_history(
    days: int = 7,
    _user: Optional[User] = Depends(get_current_user_optional),
) -> list:
    """Return historical screener runs for the last N days (Pro only, max 30).

    Each entry: {id, ran_at, results_count, top_5_tickers, error}.
    Useful for tracking which stocks the screener consistently identifies.
    """
    days = max(1, min(days, 30))

    try:
        with get_conn() as c:
            with c.cursor() as cur:
                cur.execute(
                    """SELECT id, ran_at, payload
                       FROM screener_results
                       WHERE ran_at >= now() - (%s || ' days')::interval
                       ORDER BY ran_at DESC""",
                    [str(days)],
                )
                rows = cur.fetchall()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"DB error: {e}")

    history = []
    for row_id, ran_at, payload_data in rows:
        payload = json.loads(payload_data) if isinstance(payload_data, str) else payload_data
        results = payload.get("results", [])
        top5 = [r.get("ticker") for r in results[:5]]
        history.append({
            "id": row_id,
            "ran_at": ran_at.isoformat() if ran_at else None,
            "results_count": len(results),
            "top_5_tickers": top5,
            "error": payload.get("error"),
        })

    return history


@router.get("/share/{result_id}")
def get_shared_screener(result_id: int) -> Dict[str, Any]:
    """Public read of a specific screener run by ID (no auth required).

    Powers shareable result links: /results/{id}
    """
    try:
        with get_conn() as c:
            with c.cursor() as cur:
                cur.execute(
                    "SELECT ran_at, payload FROM screener_results WHERE id = %s",
                    [result_id],
                )
                row = cur.fetchone()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"DB error: {e}")

    if not row:
        raise HTTPException(status_code=404, detail="Screener result not found.")

    ran_at, payload_data = row
    payload = json.loads(payload_data) if isinstance(payload_data, str) else payload_data
    results = payload.get("results", [])

    return {
        "id": result_id,
        "ran_at": ran_at.isoformat() if ran_at else None,
        "results_count": len(results),
        "top_picks": results[:10],
    }

File: worker/src/routes/screener.ts
// Screener routes — port of api/routes/screener.py.
//
// /cached, /top, /history, /share/:id read the cron-assembled screener_results.
// /run scores a custom ticker set live (≤ 40 to stay inside subrequest budget).
// /export streams the latest run as CSV.

import { Hono } from "hono";
import { z } from "zod";
import { marketRegime, type BenchmarkRow } from "../engine/benchmarks";
import { applyFilters, type FilterConfig } from "../engine/filters";
import { computeMetrics, type Metrics } from "../engine/metrics";
import { scoreRecords } from "../engine/orchestrator";
import { fetchChart } from "../lib/yahoo";
import type { AppContext, Env } from "../types";

// Live-run cap: each ticker costs one chart subrequest; keep headroom under
// the 50-subrequest free-plan budget.
const MAX_LIVE_TICKERS = 40;

const runSchema = z.object({
  tickers: z
    .array(z.string().transform((s) => s.trim().toUpperCase()))
    .max(MAX_LIVE_TICKERS + 1)
    .optional(),
  filters: z.record(z.unknown()).optional(),
  agents: z.array(z.string()).optional(),
});

export const screenerRoutes = new Hono<AppContext>();

async function loadBenchmarks(env: Env): Promise<Record<string, BenchmarkRow>> {
  const rows = await env.DB.prepare("SELECT * FROM benchmark_metrics").all<BenchmarkRow>();
  const out: Record<string, BenchmarkRow> = {};
  for (const b of rows.results ?? []) out[b.ticker] = b;
  return out;
}

function decorate(
  r: ReturnType<typeof scoreRecords>[number],
  benchmarks: Record<string, BenchmarkRow>,
) {
  return {
    ...r,
    score: r.composite_score,
    ret_7d: r.change_7d,
    momentum: r["score_momentum"],
    breakout: r["score_breakout"],
    volume: r["score_volume_surge"],
    rs: r["score_relative_strength"],
    mean_reversion: r["score_mean_reversion"],
    vs_voo: r.change_7d - (benchmarks["VOO"]?.change_7d ?? 0),
  };
}

screenerRoutes.post("/run", async (c) => {
  const parsed = runSchema.safeParse(await c.req.json().catch(() => ({})));
  if (!parsed.success) return c.json({ detail: "Invalid payload" }, 422);
  const { tickers, filters, agents } = parsed.data;

  if (!tickers || tickers.length === 0) {
    return c.json(
      { detail: "Provide up to 40 tickers. Full-universe results come from GET /api/screener/cached." },
      422,
    );
  }
  if (tickers.length > MAX_LIVE_TICKERS) {
    return c.json(
      {
        detail: `Live screener accepts up to ${MAX_LIVE_TICKERS} tickers. Use GET /api/screener/cached for the full universe.`,
      },
      413,
    );
  }

  const unique = [...new Set(tickers)].filter((t) => /^[A-Z][A-Z0-9.\-]{0,9}$/.test(t));
  const records: Metrics[] = [];
  for (let i = 0; i < unique.length; i += 8) {
    const batch = unique.slice(i, i + 8);
    const settled = await Promise.all(
      batch.map(async (t) => {
        const bars = await fetchChart(t, "35d");
        return bars ? computeMetrics(t, bars) : null;
      }),
    );
    records.push(...settled.filter((m): m is Metrics => m !== null));
  }

  const benchmarks = await loadBenchmarks(c.env);
  const regime = marketRegime(benchmarks);
  let scored = scoreRecords(records, { benchmarks, regime, agentNames: agents ?? null });
  if (filters && Object.keys(filters).length > 0) {
    scored = applyFilters(scored, filters as FilterConfig);
  }

  return c.json({
    regime,
    benchmarks: Object.values(benchmarks),
    results: scored.map((r) => decorate(r, benchmarks)),
    ran_at: new Date().toISOString(),
  });
});

async function latestResult(env: Env): Promise<{ ran_at: string; payload: string } | null> {
  return env.DB.prepare(
    "SELECT ran_at, payload FROM screener_results ORDER BY ran_at DESC LIMIT 1",
  ).first<{ ran_at: string; payload: string }>();
}

screenerRoutes.get("/cached", async (c) => {
  const row = await latestResult(c.env);
  if (!row) return c.json({ detail: "No cached results yet. The refresh cron will populate them." }, 404);
  const payload = JSON.parse(row.payload);
  return c.json({
    regime: payload.regime ?? {},
    benchmarks: payload.benchmarks ?? [],
    results: payload.results ?? [],
    ran_at: payload.ran_at ?? row.ran_at,
  });
});

screenerRoutes.get("/top", async (c) => {
  const n = Number(c.req.query("n") ?? 10);
  if (!Number.isInteger(n) || n < 1 || n > 100) {
    return c.json({ detail: "n must be between 1 and 100" }, 400);
  }
  const row = await latestResult(c.env);
  if (!row) return c.json([]);
  const results = (JSON.parse(row.payload).results ?? []) as Record<string, unknown>[];
  return c.json(
    results.slice(0, n).map((r) => ({
      ticker: r.ticker,
      ret_7d: r.ret_7d ?? r.change_7d,
      score: r.score ?? r.composite_score,
      best_strategy: r.best_strategy,
      vs_voo: r.vs_voo,
    })),
  );
});

screenerRoutes.get("/history", async (c) => {
  const days = Math.max(1, Math.min(Number(c.req.query("days") ?? 7) || 7, 30));
  const rows = await c.env.DB.prepare(
    `SELECT id, ran_at, payload FROM screener_results
     WHERE ran_at >= datetime('now', '-${days} days') ORDER BY ran_at DESC`,
  ).all<{ id: number; ran_at: string; payload: string }>();
  return c.json(
    (rows.results ?? []).map((row) => {
      const payload = JSON.parse(row.payload);
      const results = (payload.results ?? []) as { ticker: string }[];
      return {
        id: row.id,
        ran_at: payload.ran_at ?? row.ran_at,
        results_count: results.length,
        top_5_tickers: results.slice(0, 5).map((r) => r.ticker),
        error: payload.error ?? null,
      };
    }),
  );
});

screenerRoutes.get("/share/:id", async (c) => {
  const id = Number(c.req.param("id"));
  if (!Number.isInteger(id) || id < 1) return c.json({ detail: "Invalid id" }, 400);
  const row = await c.env.DB.prepare(
    "SELECT ran_at, payload FROM screener_results WHERE id = ?",
  )
    .bind(id)
    .first<{ ran_at: string; payload: string }>();
  if (!row) return c.json({ detail: "Screener result not found." }, 404);
  const payload = JSON.parse(row.payload);
  const results = (payload.results ?? []) as Record<string, unknown>[];
  return c.json({
    id,
    ran_at: payload.ran_at ?? row.ran_at,
    results_count: results.length,
    top_picks: results.slice(0, 10),
  });
});

screenerRoutes.get("/export", async (c) => {
  const row = await latestResult(c.env);
  if (!row) return c.json({ detail: "No screener results to export yet." }, 404);
  const payload = JSON.parse(row.payload);
  const results = (payload.results ?? []) as Record<string, unknown>[];
  if (results.length === 0) return c.json({ detail: "No results in latest screener run." }, 404);

  const lead = [
    "ticker", "price", "ret_7d", "change_7d", "score", "composite_score",
    "momentum", "breakout", "volume", "rs", "mean_reversion", "best_strategy", "vs_voo",
  ];
  const extras = Object.keys(results[0]).filter((k) => !lead.includes(k)).sort();
  const fields = [...lead, ...extras];

  const esc = (v: unknown): string => {
    if (v === null || v === undefined) return "";
    const s = String(v);
    return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  };
  const csv = [
    fields.join(","),
    ...results.map((r) => fields.map((f) => esc(r[f])).join(",")),
  ].join("\n");

  const date = (payload.ran_at ?? row.ran_at ?? "").slice(0, 10).replace(/-/g, "") || "latest";
  return c.body(csv, 200, {
    "Content-Type": "text/csv; charset=utf-8",
    "Content-Disposition": `attachment; filename="best7daysmula_screener_${date}.csv"`,
  });
});
